File: rag_backend/app/models/group_chat.py
from sqlalchemy import Column, String, DateTime, Boolean, Integer, ForeignKey, JSON, Text, Enum as SQLEnum
from sqlalchemy.orm import relationship
from app.db.base import Base
import uuid
from datetime import datetime
from enum import Enum


class GroupMemberStatus(str, Enum):
    INVITED = "invited"
    ACTIVE = "active"
    LEFT = "left"
    REMOVED = "removed"


class GroupRole(str, Enum):
    OWNER = "owner"
    ADMIN = "admin"
    MEMBER = "member"


class ChatGroup(Base):
    __tablename__ = "chat_groups"
    
    id = Column(String, primary_key=True, default=lambda: f"grp_{uuid.uuid4().hex}")
    tenant_id = Column(String, nullable=False, index=True)
    
    name = Column(String, nullable=False)
    description = Column(String)
    avatar_url = Column(String, nullable=True)
    
    status = Column(String, default="active")
    created_by = Column(String, nullable=False)
    
    settings = Column(JSON, default=lambda: {
        "allow_member_invite": True,
        "require_approval": False,
        "max_members": 100
    })
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    archived_at = Column(DateTime, nullable=True)
    
    members = relationship("GroupMember", back_populates="group", lazy="dynamic")
    messages = relationship("GroupMessage", back_populates="group", lazy="dynamic")
    invitations = relationship("GroupInvitation", back_populates="group", lazy="dynamic")
    
    def __repr__(self):
        return f"<ChatGroup(id={self.id}, name={self.name})>"


class GroupMember(Base):
    __tablename__ = "group_members"
    
    id = Column(String, primary_key=True, default=lambda: f"gmem_{uuid.uuid4().hex}")
    group_id = Column(String, ForeignKey("chat_groups.id"), nullable=False)
    
    user_id = Column(String, nullable=False)
    tenant_id = Column(String, nullable=False, index=True)
    
    role = Column(String, default=GroupRole.MEMBER.value)
    status = Column(String, default=GroupMemberStatus.ACTIVE.value)
    
    invited_by = Column(String, nullable=True)
    joined_at = Column(DateTime, default=datetime.utcnow)
    left_at = Column(DateTime, nullable=True)
    
    notification_settings = Column(JSON, default=lambda: {
        "all_messages": True,
        "mentions_only": False
    })
    
    group = relationship("ChatGroup", back_populates="members")
    
    def __repr__(self):
        return f"<GroupMember(id={self.id}, user_id={self.user_id}, group_id={self.group_id})>"


class GroupInvitation(Base):
    __tablename__ = "group_invitations"
    
    id = Column(String, primary_key=True, default=lambda: f"ginv_{uuid.uuid4().hex}")
    group_id = Column(String, ForeignKey("chat_groups.id"), nullable=False)
    
    invitee_id = Column(String, nullable=False)
    inviter_id = Column(String, nullable=False)
    tenant_id = Column(String, nullable=False, index=True)
    
    status = Column(String, default="pending")
    message = Column(String, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    expires_at = Column(DateTime, nullable=True)
    responded_at = Column(DateTime, nullable=True)
    
    group = relationship("ChatGroup", back_populates="invitations")
    
    def __repr__(self):
        return f"<GroupInvitation(id={self.id}, invitee_id={self.invitee_id}, status={self.status})>"


class GroupMessage(Base):
    __tablename__ = "group_messages"
    
    id = Column(String, primary_key=True, default=lambda: f"gmsg_{uuid.uuid4().hex}")
    group_id = Column(String, ForeignKey("chat_groups.id"), nullable=False)
    
    sender_id = Column(String, nullable=False)
    tenant_id = Column(String, nullable=False, index=True)
    
    content = Column(Text, nullable=False)
    content_type = Column(String, default="text")
    
    metadata_ = Column("metadata", JSON, default=dict)
    
    is_deleted = Column(Boolean, default=False)
    is_edited = Column(Boolean, default=False)
    edited_at = Column(DateTime, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    
    group = relationship("ChatGroup", back_populates="messages")
    
    def __repr__(self):
        return f"<GroupMessage(id={self.id}, sender_id={self.sender_id})>"
    
    def to_dict(self):
        return {
            "id": self.id,
            "group_id": self.group_id,
            "sender_id": self.sender_id,
            "tenant_id": self.tenant_id,
            "content": self.content,
            "content_type": self.content_type,
            "metadata": self.metadata_,
            "is_deleted": self.is_deleted,
            "is_edited": self.is_edited,
            "edited_at": self.edited_at.isoformat() if self.edited_at else None,
            "created_at": self.created_at.isoformat()
        }

File: rag_frontend/src/api/group-chat.ts
import { request } from '@/utils/request'

export interface ChatGroup {
  id: string
  tenant_id: string
  name: string
  description?: string
  avatar_url?: string
  status: string
  created_by: string
  created_at: string
  updated_at: string
  member_count?: number
  last_message?: GroupMessage
}

export interface GroupMember {
  id: string
  group_id: string
  user_id: string
  user_name?: string
  avatar_url?: string
  tenant_id: string
  role: 'owner' | 'admin' | 'member'
  status: 'invited' | 'active' | 'left' | 'removed'
  joined_at?: string
  notification_settings?: {
    enabled: boolean
    mentions_only: boolean
  }
}

export interface GroupInvitation {
  id: string
  group_id: string
  group_name?: string
  invitee_id: string
  inviter_id: string
  inviter_name?: string
  tenant_id: string
  status: 'pending' | 'accepted' | 'declined' | 'expired'
  message?: string
  created_at: string
  expires_at?: string
}

export interface GroupMessage {
  id: string
  group_id: string
  sender_id: string
  sender_name?: string
  sender_avatar?: string
  tenant_id: string
  content: string
  content_type: 'text' | 'image' | 'file' | 'system'
  metadata?: Record<string, any>
  created_at: string
  is_deleted?: boolean
}

export interface Notification {
  id: string
  type: 'invitation' | 'message' | 'member_joined' | 'member_left' | 'system'
  title: string
  content: string
  group_id?: string
  group_name?: string
  is_read: boolean
  created_at: string
  data?: Record<string, any>
}

export interface CreateGroupRequest {
  name: string
  description?: string
  avatar_url?: string
}

export interface InviteMembersRequest {
  invitee_ids: string[]
  message?: string
}

export interface SendMessageRequest {
  content: string
  content_type?: 'text' | 'image' | 'file'
}

export const groupChatApi = {
  async getGroups(): Promise<ChatGroup[]> {
    return request.get('/groups/')
  },

  async createGroup(data: CreateGroupRequest): Promise<ChatGroup> {
    return request.post('/groups/', data)
  },

  async getGroup(groupId: string): Promise<ChatGroup> {
    return request.get(`/groups/${groupId}`)
  },

  async updateGroup(groupId: string, data: Partial<CreateGroupRequest>): Promise<ChatGroup> {
    return request.put(`/groups/${groupId}`, data)
  },

  async deleteGroup(groupId: string): Promise<void> {
    return request.delete(`/groups/${groupId}`)
  },

  async getGroupMembers(groupId: string): Promise<GroupMember[]> {
    return request.get(`/groups/${groupId}/members`)
  },

  async inviteMembers(groupId: string, data: InviteMembersRequest): Promise<GroupInvitation[]> {
    return request.post(`/groups/${groupId}/invite`, data)
  },

  async leaveGroup(groupId: string): Promise<void> {
    return request.post(`/groups/${groupId}/leave`)
  },

  async removeMember(groupId: string, userId: string): Promise<void> {
    return request.delete(`/groups/${groupId}/members/${userId}`)
  },

  async updateMemberRole(groupId: string, userId: string, role: 'admin' | 'member'): Promise<GroupMember> {
    return request.put(`/groups/${groupId}/members/${userId}`, { role })
  },

  async getGroupMessages(groupId: string, params?: { limit?: number; before?: string }): Promise<GroupMessage[]> {
    return request.get(`/groups/${groupId}/messages`, { params })
  },

  async sendMessage(groupId: string, data: SendMessageRequest): Promise<GroupMessage> {
    return request.post(`/groups/${groupId}/messages`, data)
  },

  async getPendingInvitations(): Promise<GroupInvitation[]> {
    return request.get('/invitations/pending')
  },

  async acceptInvitation(invitationId: string): Promise<GroupMember> {
    return request.post(`/invitations/${invitationId}/accept`)
  },

  async declineInvitation(invitationId: string): Promise<void> {
    return request.post(`/invitations/${invitationId}/decline`)
  },

  async getNotifications(params?: { limit?: number; unread_only?: boolean }): Promise<Notification[]> {
    return request.get('/notifications/', { params })
  },

  async markNotificationRead(notificationId: string): Promise<void> {
    return request.put(`/notifications/${notificationId}/read`)
  },

  async markAllNotificationsRead(): Promise<void> {
    return request.put('/notifications/read-all')
  },

  async deleteNotification(notificationId: string): Promise<void> {
    return request.delete(`/notifications/${notificationId}`)
  },

  async deleteNotificationsBatch(notificationIds: string[]): Promise<{ deleted_count: number }> {
    return request.post('/notifications/delete-batch', notificationIds)
  },

  async clearAllNotifications(): Promise<void> {
    return request.delete('/notifications/')
  },

  async getOnlineMembers(groupId: string): Promise<string[]> {
    return request.get(`/groups/${groupId}/online-members`)
  },

  createWebSocketConnection(groupId: string): WebSocket {
    const token = localStorage.getItem('rag_token')
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
    const wsUrl = `${protocol}//${window.location.host}/api/v1/ws/groups/${groupId}?token=${token}`
    return new WebSocket(wsUrl)
  }
}

export interface WebSocketMessage {
  type: 'new_message' | 'member_joined' | 'member_left' | 'member_removed' | 
        'group_updated' | 'typing' | 'online_status' | 'invitation_received' |
        'notification' | 'error'
  data?: any
  sender_id?: string
  timestamp?: string
}
